src/ProbabilisticProgramming/Monad.dfy

```diff
@@ -57,18 +107,50 @@ module Monad {
       case Diverging => false
       case Result(_, rest) => property(rest)
     }
+
+    predicate IsConverging() {
+      Result?
+    }
+
+    predicate IsDiverging() {
+      Diverging?
+    }
+
+    function ValueOf(): A
+      requires this.IsConverging()
+    {
+      this.value
+    }
+
+    function RestOf(): Rand.Bitstream
+      requires this.IsConverging()
+    {
+      this.rest
+    }
+  }
+
+  function ConvergingResult<A>(value: A, rest: Rand.Bitstream): Result<A> {
```

Comment:
This name feels too long. How about `Terminating` or `TerminatingWith`? Another idea is to rename the `Result` variant to `Terminating`, so that this function can be called `Result`.